shared/vendored/snakemake/versioning.smk

```diff
@@ -0,0 +1,163 @@
+import subprocess
+import re
+
+class DependencyChecker():
+    def __init__(self, registration):
+        super().__init__()
+        self.declared_dependencies = registration.get('dependencies', {})
+        self.error_attrs = ["version_incompatibilities", "not_found_dependencies", "declaration_errors", "unexpected_errors"]
+        for attr in self.error_attrs:
+            setattr(self, attr, [])
+
+    def check(self):
+        for package_name, declared_spec in self.declared_dependencies.get('py', {}).items():
+            self.check_python_package(package_name, declared_spec)
+        for program_name, declared_spec in self.declared_dependencies.get('cli', {}).items():
+            self.check_cli_version(program_name, declared_spec)
+
+    def report_errors(self) -> bool:
+        if sum([len(getattr(self, attr)) for attr in self.error_attrs])==0:
+            print("All dependencies declared by this pathogen satisfied", file=stderr)
+            return False
+        
+        print(f"This pathogen declares dependencies which were not met.", file=stderr)
+        for attr in self.error_attrs:
+            errors = getattr(self, attr)
+            if len(errors)==0:
+                continue
+            print(attr.replace('_', ' ').upper() + ":")
+            for msg in errors:
+                print(f"\t{msg}", file=stderr)
+        return True
+
+    def check_python_package(self, name: str, declared_spec: str):
+        """
+        Check whether the installed python library *name* meets the specifier *spec_str*.
+        This uses importlib.metadata to check the available version which avoids importing
+        the top-level import. Any errors or if the declared dependency isn't satisfied are
+        logged.
+        """
+        try:
+            if name=='snakemake':
+                # importlib reports 0.0.0, so follow the approach of Snakemake's own min_version function
+                version = Version(snakemake_version)
+            else:
+                version = Version(importlib_version(name))
+        except PackageNotFoundError:
+            self.not_found_dependencies.append(f"Python dependency {name!r} was not found. This pathogen requires a version satisfying {declared_spec}")
+            return
+        except InvalidVersion: # <https://packaging.pypa.io/en/stable/version.html#packaging.version.InvalidVersion>
+            self.unexpected_errors.append(f"Python dependency {name!r} reported a version of {output} which we were unable to parse")
+            return
+
+        try:
+            spec = SpecifierSet(declared_spec)
+        except InvalidSpecifier:
+            # TODO XXX: I declared version '!31' which didn't raise an InvalidSpecifier but also didn't work as expected
+            # as installed version 31.3.0 was acceptable. '!31.*' or '!31.3' or '!31.3.0' failed.
+            self.declaration_errors.append(f"This pathogen declared an invalid version specification for {name!r} of {declared_spec}")
```

Comment:
I think helpful to have a hint here of what are the expected version specification formats. 
Maybe https://packaging.python.org/en/latest/specifications/version-specifiers/#id5?